### ISSUE TITLE
How to customize a new loss to receive multiple arguments from a new header?

### ISSUE THREAD

**Issue by Cogito2012** [open-mmlab/mmaction2#440]
The provided example in the API doc [Add New Loss](https://mmaction2.readthedocs.io/en/latest/tutorials/new_modules.html#add-new-loss) is too simple and seems not applicable for complex arguments. For example:

Here is a new customized header:
```python
@HEADS.register_module()
class MyHead(BaseHead):
    def __init__(self, arg1, arg2):
        pass

    def forward(self, x):
        # the forward pass generate multiple outputs (cls_score, B, C)
        return cls_score, B, C
```

And here is the corresponding new loss:
```python
@LOSSES.register_module()
class MyLoss(nn.Module):

    def forward(self, cls_score, target, B, C):
        loss_1 = my_loss1(cls_score, target)
        loss_2 = my_loss2(B, C)
        loss = loss1 + loss2
        return loss
    
    def my_loss1(pred, target):
        # implement loss1
    def my_loss2(B, C):
        # implement loss2
```

Is there any way to feed the additional outputs `B` and `C` into the forward function of the new loss? As a beginner to use mmaction2, I am not clear about how the loss module receives arguments from the header module. I appreciate any help or hint to do that. Thanks!

**Comment by innerlee** [comment]
Many losses have customized arguments, see eg. https://github.com/open-mmlab/mmaction2/blob/master/mmaction/models/losses/binary_logistic_regression_loss.py#L39-L44, https://github.com/open-mmlab/mmaction2/blob/master/mmaction/models/losses/bmn_loss.py#L134-L144, https://github.com/open-mmlab/mmaction2/blob/master/mmaction/models/losses/hvu_loss.py#L61

**Comment by irvingzhang0512** [comment]
What a coincidence! I met the same problem yesterday...Yes, we can set additional outputs from models. 

+ Two steps you need to implement this: 
    1. Add new keys in input batch. `{'imgs': ..., 'label': ..., 'B': ..., 'C': ...}`
    2. set these new keys in `train_cfg['aux_info'] =['B', 'c']`

**Comment by innerlee** [comment]
The extra arguments are passed here https://github.com/open-mmlab/mmaction2/blob/5e0ffc1cd695452eae168d82706b2c6f92a868da/mmaction/models/recognizers/recognizer2d.py#L30

**Comment by Cogito2012** [comment]
> The extra arguments are passed here
> 
> https://github.com/open-mmlab/mmaction2/blob/5e0ffc1cd695452eae168d82706b2c6f92a868da/mmaction/models/recognizers/recognizer2d.py#L30

Thanks for this hint! I rewrote the `mmaction/models/recognizers/recognizer2d.py` so that the extra outputs from the header can be fed into the new loss now. This hint really helps a lot!

@irvingzhang0512  Right, you got my point! I do not have extra info from data, but extra outputs from the model. Your method is a good solution to add extra inputs from data. Thanks for this information.

**Comment by innerlee** [comment]
the `**kwargs` collects arbitrary keyword arguments, so I think maybe you do not have to rewrite it

**Comment by Cogito2012** [comment]
> the `**kwargs` collects arbitrary keyword arguments, so I think maybe you do not have to rewrite it

Yeah, ideally it should be better as you said. However, in the `recognizer2d.py`, the `self.cls_head` returns only one output `cls_score`, see [L28 in recognizer2d.py](https://github.com/open-mmlab/mmaction2/blob/5e0ffc1cd695452eae168d82706b2c6f92a868da/mmaction/models/recognizers/recognizer2d.py#L28):

```python
        cls_score = self.cls_head(x, num_segs)  # Here, I have multiple outputs, not a single cls_score.
        gt_labels = labels.squeeze()
        loss_cls = self.cls_head.loss(cls_score, gt_labels, **kwargs)  # Here,  **kwargs cannot collects multiple outputs from self.cls_head()
```

So, I have to change this to:
```python
        cls_score, B, C = self.cls_head(x, num_segs)  # Here, I have multiple outputs, not a single cls_score.
        gt_labels = labels.squeeze()
        loss_cls = self.cls_head.my_loss(cls_score, gt_labels, B, C, **kwargs)  # Here, the `my_loss` is instantiated by build_loss in the class MyHead.
```

Admittedly, this is an awkward way to solve the problem. I would appreciate it if you have a better approach to this issue.

**Comment by innerlee** [comment]
I see, you want to return multiple outputs for the head.

How about writing a customized Head that returns a tuple of `(cls_score, B, C)`?

**Comment by Cogito2012** [comment]
> I see, you want to return multiple outputs for the head.
> 
> How about writing a customized Head that returns a tuple of `(cls_score, B, C)`?

Right, that's what I have done. But it is still necessary to modify `recognizer2d.py`, otherwise it will at least causes error in [Line 66 and 68](https://github.com/open-mmlab/mmaction2/blob/5e0ffc1cd695452eae168d82706b2c6f92a868da/mmaction/models/recognizers/recognizer2d.py#L68) in `recognizer2d.py`.

So, in addition to returns a tuple for the customized Head, I think a modified `recognizer2d.py` is still needed.

**Comment by innerlee** [comment]
Okay, how about write a new MyRecognizer2D class inherit from Recognizer2D

**Comment by Cogito2012** [comment]
@innerlee Yes, it's better to create a new class to inherit Recognizer2D and only override the methods that need to be changed.

**Comment by Better369** [comment]
> @innerlee Yes, it's better to create a new class to inherit Recognizer2D and only override the methods that need to be changed.

Hello, I would like to achieve the same functionality as you. Could you please provide me with your contact information for a detailed exchange?
